File: frontend/services/validation.test.ts
/**
 * Frontend validation tests
 * These tests ensure the frontend validation matches backend validation
 */

import {
  validatePersonaggio,
  validateProductCreate,
  validateProductUpdate,
  validateVariant,
  validateProductImage,
  validateImageFile,
  Validator,
} from './validation';

describe('Personaggio Validation', () => {
  it('should validate valid personaggio with all fields', () => {
    const data = {
      name: 'Test Character',
      description: 'A test character description',
      icon: '/uploads/icon.png',
      images: ['/uploads/img1.png', '/uploads/img2.png'],
      backgroundColor: '#FF0000',
      backgroundType: 'solid' as const,
      order: 0,
    };

    const result = validatePersonaggio(data);
    expect(result.hasErrors()).toBe(false);
  });

  it('should validate valid personaggio with gradient', () => {
    const data = {
      name: 'Gradient Character',
      description: 'Character with gradient background',
      images: [],
      backgroundColor: '#FF0000',
      backgroundType: 'gradient' as const,
      gradientFrom: '#FF0000',
      gradientTo: '#0000FF',
      order: 1,
    };

    const result = validatePersonaggio(data);
    expect(result.hasErrors()).toBe(false);
  });

  it('should validate valid personaggio with minimal fields', () => {
    const data = {
      name: 'Minimal Character',
      images: [],
    };

    const result = validatePersonaggio(data);
    expect(result.hasErrors()).toBe(false);
  });

  it('should reject empty name', () => {
    const data = {
      name: '',
      images: [],
    };

    const result = validatePersonaggio(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('name')).toBeDefined();
  });

  it('should reject name too long', () => {
    const data = {
      name: 'This is a very long name that exceeds the maximum allowed length of 100 characters for a personaggio name field',
      images: [],
    };

    const result = validatePersonaggio(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('name')).toBeDefined();
  });

  it('should reject too many images', () => {
    const data = {
      name: 'Valid Name',
      images: Array(21).fill('/img.png'), // 21 images - exceeds limit
    };

    const result = validatePersonaggio(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('images')).toBeDefined();
  });

  it('should reject bad hex color', () => {
    const data = {
      name: 'Valid Name',
      images: [],
      backgroundColor: 'red', // Not a hex color
    };

    const result = validatePersonaggio(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('backgroundColor')).toBeDefined();
  });

  it('should reject invalid background type', () => {
    const data = {
      name: 'Valid Name',
      images: [],
      backgroundType: 'pattern' as any, // Invalid type
    };

    const result = validatePersonaggio(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('backgroundType')).toBeDefined();
  });

  it('should reject negative order', () => {
    const data = {
      name: 'Valid Name',
      images: [],
      order: -1,
    };

    const result = validatePersonaggio(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('order')).toBeDefined();
  });

  it('should reject bad URL format', () => {
    const data = {
      name: 'Valid Name',
      icon: 'not-a-url',
      images: [],
    };

    const result = validatePersonaggio(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('icon')).toBeDefined();
  });
});

describe('Product Create Validation', () => {
  it('should validate valid product', () => {
    const data = {
      title: 'Test Product',
      slug: 'test-product',
      short_description: 'A test product',
      base_price: 99.99,
      currency: 'EUR',
      sku: 'TEST-001',
      status: 'draft' as const,
    };

    const result = validateProductCreate(data);
    expect(result.hasErrors()).toBe(false);
  });

  it('should validate valid product with minimal fields', () => {
    const data = {
      title: 'Minimal Product',
      slug: 'minimal-product',
      base_price: 0,
    };

    const result = validateProductCreate(data);
    expect(result.hasErrors()).toBe(false);
  });

  it('should reject empty title', () => {
    const data = {
      title: '',
      slug: 'test-product',
      base_price: 10.0,
    };

    const result = validateProductCreate(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('title')).toBeDefined();
  });

  it('should reject empty slug', () => {
    const data = {
      title: 'Test Product',
      slug: '',
      base_price: 10.0,
    };

    const result = validateProductCreate(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('slug')).toBeDefined();
  });

  it('should reject slug with uppercase', () => {
    const data = {
      title: 'Test Product',
      slug: 'Test-Product',
      base_price: 10.0,
    };

    const result = validateProductCreate(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('slug')).toBeDefined();
  });

  it('should reject slug with spaces', () => {
    const data = {
      title: 'Test Product',
      slug: 'test product',
      base_price: 10.0,
    };

    const result = validateProductCreate(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('slug')).toBeDefined();
  });

  it('should reject negative price', () => {
    const data = {
      title: 'Test Product',
      slug: 'test-product',
      base_price: -10.0,
    };

    const result = validateProductCreate(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('base_price')).toBeDefined();
  });

  it('should reject invalid status', () => {
    const data = {
      title: 'Test Product',
      slug: 'test-product',
      base_price: 10.0,
      status: 'invalid' as any,
    };

    const result = validateProductCreate(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('status')).toBeDefined();
  });
});

describe('Product Update Validation', () => {
  it('should validate title update', () => {
    const data = {
      title: 'Updated Title',
    };

    const result = validateProductUpdate(data);
    expect(result.hasErrors()).toBe(false);
  });

  it('should validate price update', () => {
    const data = {
      base_price: 199.99,
    };

    const result = validateProductUpdate(data);
    expect(result.hasErrors()).toBe(false);
  });

  it('should validate full update', () => {
    const data = {
      title: 'Updated Product',
      slug: 'updated-product',
      short_description: 'Updated description',
      base_price: 150.0,
      status: 'published' as const,
    };

    const result = validateProductUpdate(data);
    expect(result.hasErrors()).toBe(false);
  });

  it('should reject slug with uppercase', () => {
    const data = {
      slug: 'Invalid-Slug',
    };

    const result = validateProductUpdate(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('slug')).toBeDefined();
  });

  it('should reject negative price', () => {
    const data = {
      base_price: -50.0,
    };

    const result = validateProductUpdate(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('base_price')).toBeDefined();
  });
});

describe('Variant Validation', () => {
  it('should validate valid variant', () => {
    const data = {
      sku: 'VAR-001',
      name: 'Size M',
      attributes: '{"size": "M"}',
      price_adjustment: 5.0,
      stock: 100,
    };

    const result = validateVariant(data);
    expect(result.hasErrors()).toBe(false);
  });

  it('should validate minimal variant', () => {
    const data = {
      sku: 'VAR-002',
      name: 'Default',
      stock: 0,
    };

    const result = validateVariant(data);
    expect(result.hasErrors()).toBe(false);
  });

  it('should reject empty SKU', () => {
    const data = {
      sku: '',
      name: 'Size M',
      stock: 10,
    };

    const result = validateVariant(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('sku')).toBeDefined();
  });

  it('should reject empty name', () => {
    const data = {
      sku: 'VAR-003',
      name: '',
      stock: 10,
    };

    const result = validateVariant(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('name')).toBeDefined();
  });

  it('should reject negative stock', () => {
    const data = {
      sku: 'VAR-004',
      name: 'Size L',
      stock: -5,
    };

    const result = validateVariant(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('stock')).toBeDefined();
  });
});

describe('Product Image Validation', () => {
  it('should validate valid image', () => {
    const data = {
      url: '/uploads/product-image.png',
      alt_text: 'Product image',
      position: 0,
    };

    const result = validateProductImage(data);
    expect(result.hasErrors()).toBe(false);
  });

  it('should validate image with https URL', () => {
    const data = {
      url: 'https://example.com/image.png',
      position: 1,
    };

    const result = validateProductImage(data);
    expect(result.hasErrors()).toBe(false);
  });

  it('should reject empty URL', () => {
    const data = {
      url: '',
      position: 0,
    };

    const result = validateProductImage(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('url')).toBeDefined();
  });

  it('should reject bad URL format', () => {
    const data = {
      url: 'not-a-url',
      position: 0,
    };

    const result = validateProductImage(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('url')).toBeDefined();
  });

  it('should reject negative position', () => {
    const data = {
      url: '/uploads/image.png',
      position: -1,
    };

    const result = validateProductImage(data);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('position')).toBeDefined();
  });
});

describe('Image File Validation', () => {
  it('should validate valid image file', () => {
    const file = new File([''], 'test.png', { type: 'image/png' });
    Object.defineProperty(file, 'size', { value: 1024 * 1024 }); // 1MB

    const result = validateImageFile(file, 10);
    expect(result.hasErrors()).toBe(false);
  });

  it('should reject non-image file', () => {
    const file = new File([''], 'test.pdf', { type: 'application/pdf' });

    const result = validateImageFile(file, 10);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('file')).toBeDefined();
  });

  it('should reject file too large', () => {
    const file = new File([''], 'test.png', { type: 'image/png' });
    Object.defineProperty(file, 'size', { value: 11 * 1024 * 1024 }); // 11MB

    const result = validateImageFile(file, 10);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('file')).toBeDefined();
  });

  it('should reject invalid file extension', () => {
    const file = new File([''], 'test.bmp', { type: 'image/bmp' });

    const result = validateImageFile(file, 10);
    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('file')).toBeDefined();
  });
});

describe('Validator Class', () => {
  it('should validate required fields', () => {
    const validator = new Validator();
    validator.required('test', '');
    const result = validator.getResult();

    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('test')).toBeDefined();
  });

  it('should validate minimum length', () => {
    const validator = new Validator();
    validator.minLength('test', 'ab', 3);
    const result = validator.getResult();

    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('test')).toBeDefined();
  });

  it('should validate maximum length', () => {
    const validator = new Validator();
    validator.maxLength('test', 'abcdef', 5);
    const result = validator.getResult();

    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('test')).toBeDefined();
  });

  it('should validate minimum value', () => {
    const validator = new Validator();
    validator.minValue('price', 5.0, 10.0);
    const result = validator.getResult();

    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('price')).toBeDefined();
  });

  it('should validate maximum value', () => {
    const validator = new Validator();
    validator.maxValue('price', 150.0, 100.0);
    const result = validator.getResult();

    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('price')).toBeDefined();
  });

  it('should validate email format', () => {
    const validator = new Validator();
    validator.email('email', 'not-an-email');
    const result = validator.getResult();

    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('email')).toBeDefined();
  });

  it('should validate hex color format', () => {
    const validator = new Validator();
    validator.colorHex('color', 'red');
    const result = validator.getResult();

    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('color')).toBeDefined();
  });

  it('should accept valid hex color', () => {
    const validator = new Validator();
    validator.colorHex('color', '#FF0000');
    const result = validator.getResult();

    expect(result.hasErrors()).toBe(false);
  });

  it('should validate oneOf constraint', () => {
    const validator = new Validator();
    validator.oneOf('status', 'invalid', ['draft', 'published', 'archived']);
    const result = validator.getResult();

    expect(result.hasErrors()).toBe(true);
    expect(result.getFieldError('status')).toBeDefined();
  });
});

File: backend/models/validation_test.go
package models

import (
	"testing"
)

func TestPersonaggioInputValidation(t *testing.T) {
	tests := []struct {
		name    string
		input   PersonaggioInput
		wantErr bool
	}{
		{
			name: "valid personaggio with all fields",
			input: PersonaggioInput{
				Name:            "Test Character",
				Description:     "A test character description",
				Icon:            "/uploads/icon.png",
				Images:          []string{"/uploads/img1.png", "/uploads/img2.png"},
				BackgroundColor: "#FF0000",
				BackgroundType:  "solid",
				Order:           0,
			},
			wantErr: false,
		},
		{
			name: "valid personaggio with gradient",
			input: PersonaggioInput{
				Name:            "Gradient Character",
				Description:     "Character with gradient background",
				Images:          []string{},
				BackgroundColor: "#FF0000",
				BackgroundType:  "gradient",
				GradientFrom:    "#FF0000",
				GradientTo:      "#0000FF",
				Order:           1,
			},
			wantErr: false,
		},
		{
			name: "valid personaggio minimal fields",
			input: PersonaggioInput{
				Name:   "Minimal Character",
				Images: []string{},
			},
			wantErr: false,
		},
		{
			name: "invalid - empty name",
			input: PersonaggioInput{
				Name:   "",
				Images: []string{},
			},
			wantErr: true,
		},
		{
			name: "invalid - name too long",
			input: PersonaggioInput{
				Name:   "This is a very long name that exceeds the maximum allowed length of 100 characters for a personaggio name field",
				Images: []string{},
			},
			wantErr: true,
		},
		{
			name: "invalid - description too long",
			input: PersonaggioInput{
				Name:        "Valid Name",
				Description: string(make([]byte, 2001)),
				Images:      []string{},
			},
			wantErr: true,
		},
		{
			name: "invalid - too many images",
			input: PersonaggioInput{
				Name: "Valid Name",
				Images: []string{
					"/img1.png", "/img2.png", "/img3.png", "/img4.png", "/img5.png",
					"/img6.png", "/img7.png", "/img8.png", "/img9.png", "/img10.png",
					"/img11.png", "/img12.png", "/img13.png", "/img14.png", "/img15.png",
					"/img16.png", "/img17.png", "/img18.png", "/img19.png", "/img20.png",
					"/img21.png", // 21 images - exceeds limit
				},
			},
			wantErr: true,
		},
		{
			name: "invalid - bad hex color",
			input: PersonaggioInput{
				Name:            "Valid Name",
				Images:          []string{},
				BackgroundColor: "red", // Not a hex color
			},
			wantErr: true,
		},
		{
			name: "invalid - bad background type",
			input: PersonaggioInput{
				Name:           "Valid Name",
				Images:         []string{},
				BackgroundType: "pattern", // Invalid type
			},
			wantErr: true,
		},
		{
			name: "invalid - gradient without colors",
			input: PersonaggioInput{
				Name:           "Valid Name",
				Images:         []string{},
				BackgroundType: "gradient",
				// Missing GradientFrom and GradientTo
			},
			wantErr: false, // Gradient colors are optional even when type is gradient
		},
		{
			name: "invalid - negative order",
			input: PersonaggioInput{
				Name:   "Valid Name",
				Images: []string{},
				Order:  -1,
			},
			wantErr: true,
		},
		{
			name: "invalid - bad URL format",
			input: PersonaggioInput{
				Name:   "Valid Name",
				Icon:   "not-a-url",
				Images: []string{},
			},
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := tt.input.Validate()
			if (err != nil) != tt.wantErr {
				t.Errorf("PersonaggioInput.Validate() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func TestValidateProductCreate(t *testing.T) {
	tests := []struct {
		name    string
		product EnhancedProduct
		wantErr bool
	}{
		{
			name: "valid product",
			product: EnhancedProduct{
				Title:            "Test Product",
				Slug:             "test-product",
				ShortDescription: "A test product",
				BasePrice:        99.99,
				Currency:         "EUR",
				SKU:              "TEST-001",
				Status:           ProductStatusDraft,
			},
			wantErr: false,
		},
		{
			name: "valid product minimal fields",
			product: EnhancedProduct{
				Title:     "Minimal Product",
				Slug:      "minimal-product",
				BasePrice: 0,
			},
			wantErr: false,
		},
		{
			name: "invalid - empty title",
			product: EnhancedProduct{
				Title:     "",
				Slug:      "test-product",
				BasePrice: 10.0,
			},
			wantErr: true,
		},
		{
			name: "invalid - empty slug",
			product: EnhancedProduct{
				Title:     "Test Product",
				Slug:      "",
				BasePrice: 10.0,
			},
			wantErr: true,
		},
		{
			name: "invalid - slug with uppercase",
			product: EnhancedProduct{
				Title:     "Test Product",
				Slug:      "Test-Product",
				BasePrice: 10.0,
			},
			wantErr: true,
		},
		{
			name: "invalid - slug with spaces",
			product: EnhancedProduct{
				Title:     "Test Product",
				Slug:      "test product",
				BasePrice: 10.0,
			},
			wantErr: true,
		},
		{
			name: "invalid - negative price",
			product: EnhancedProduct{
				Title:     "Test Product",
				Slug:      "test-product",
				BasePrice: -10.0,
			},
			wantErr: true,
		},
		{
			name: "invalid - title too long",
			product: EnhancedProduct{
				Title:     string(make([]byte, 501)),
				Slug:      "test-product",
				BasePrice: 10.0,
			},
			wantErr: true,
		},
		{
			name: "invalid - invalid status",
			product: EnhancedProduct{
				Title:     "Test Product",
				Slug:      "test-product",
				BasePrice: 10.0,
				Status:    ProductStatus("invalid"),
			},
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := ValidateProductCreate(&tt.product)
			if (err != nil) != tt.wantErr {
				t.Errorf("ValidateProductCreate() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func TestValidateProductUpdate(t *testing.T) {
	tests := []struct {
		name    string
		product EnhancedProduct
		wantErr bool
	}{
		{
			name: "valid update - title only",
			product: EnhancedProduct{
				Title: "Updated Title",
			},
			wantErr: false,
		},
		{
			name: "valid update - price only",
			product: EnhancedProduct{
				BasePrice: 199.99,
			},
			wantErr: false,
		},
		{
			name: "valid update - all fields",
			product: EnhancedProduct{
				Title:            "Updated Product",
				Slug:             "updated-product",
				ShortDescription: "Updated description",
				BasePrice:        150.0,
				Status:           ProductStatusPublished,
			},
			wantErr: false,
		},
		{
			name: "valid update - empty title means no update",
			product: EnhancedProduct{
				Title: "",
			},
			wantErr: false,
		},
		{
			name: "invalid - slug with uppercase",
			product: EnhancedProduct{
				Slug: "Invalid-Slug",
			},
			wantErr: true,
		},
		{
			name: "invalid - negative price",
			product: EnhancedProduct{
				BasePrice: -50.0,
			},
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := ValidateProductUpdate(&tt.product)
			if (err != nil) != tt.wantErr {
				t.Errorf("ValidateProductUpdate() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func TestValidateVariant(t *testing.T) {
	tests := []struct {
		name    string
		variant ProductVariant
		wantErr bool
	}{
		{
			name: "valid variant",
			variant: ProductVariant{
				SKU:             "VAR-001",
				Name:            "Size M",
				Attributes:      `{"size": "M"}`,
				PriceAdjustment: 5.0,
				Stock:           100,
			},
			wantErr: false,
		},
		{
			name: "valid variant minimal",
			variant: ProductVariant{
				SKU:   "VAR-002",
				Name:  "Default",
				Stock: 0,
			},
			wantErr: false,
		},
		{
			name: "invalid - empty SKU",
			variant: ProductVariant{
				SKU:   "",
				Name:  "Size M",
				Stock: 10,
			},
			wantErr: true,
		},
		{
			name: "invalid - empty name",
			variant: ProductVariant{
				SKU:   "VAR-003",
				Name:  "",
				Stock: 10,
			},
			wantErr: true,
		},
		{
			name: "invalid - negative stock",
			variant: ProductVariant{
				SKU:   "VAR-004",
				Name:  "Size L",
				Stock: -5,
			},
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := ValidateVariant(&tt.variant)
			if (err != nil) != tt.wantErr {
				t.Errorf("ValidateVariant() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func TestValidateProductImage(t *testing.T) {
	tests := []struct {
		name    string
		image   ProductImage
		wantErr bool
	}{
		{
			name: "valid image",
			image: ProductImage{
				URL:      "/uploads/product-image.png",
				AltText:  "Product image",
				Position: 0,
			},
			wantErr: false,
		},
		{
			name: "valid image with https URL",
			image: ProductImage{
				URL:      "https://example.com/image.png",
				Position: 1,
			},
			wantErr: false,
		},
		{
			name: "invalid - empty URL",
			image: ProductImage{
				URL:      "",
				Position: 0,
			},
			wantErr: true,
		},
		{
			name: "invalid - bad URL format",
			image: ProductImage{
				URL:      "not-a-url",
				Position: 0,
			},
			wantErr: true,
		},
		{
			name: "invalid - negative position",
			image: ProductImage{
				URL:      "/uploads/image.png",
				Position: -1,
			},
			wantErr: true,
		},
		{
			name: "invalid - alt text too long",
			image: ProductImage{
				URL:      "/uploads/image.png",
				AltText:  string(make([]byte, 501)),
				Position: 0,
			},
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := ValidateProductImage(&tt.image)
			if (err != nil) != tt.wantErr {
				t.Errorf("ValidateProductImage() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func TestValidator(t *testing.T) {
	t.Run("Required validation", func(t *testing.T) {
		v := NewValidator()
		v.Required("test", "")
		if !v.HasErrors() {
			t.Error("Expected error for empty required field")
		}
	})

	t.Run("MinLength validation", func(t *testing.T) {
		v := NewValidator()
		v.MinLength("test", "ab", 3)
		if !v.HasErrors() {
			t.Error("Expected error for string too short")
		}
	})

	t.Run("MaxLength validation", func(t *testing.T) {
		v := NewValidator()
		v.MaxLength("test", "abcdef", 5)
		if !v.HasErrors() {
			t.Error("Expected error for string too long")
		}
	})

	t.Run("MinValue validation", func(t *testing.T) {
		v := NewValidator()
		v.MinValue("price", 5.0, 10.0)
		if !v.HasErrors() {
			t.Error("Expected error for value too small")
		}
	})

	t.Run("MaxValue validation", func(t *testing.T) {
		v := NewValidator()
		v.MaxValue("price", 150.0, 100.0)
		if !v.HasErrors() {
			t.Error("Expected error for value too large")
		}
	})

	t.Run("Email validation", func(t *testing.T) {
		v := NewValidator()
		v.Email("email", "not-an-email")
		if !v.HasErrors() {
			t.Error("Expected error for invalid email")
		}
	})

	t.Run("ColorHex validation", func(t *testing.T) {
		v := NewValidator()
		v.ColorHex("color", "red")
		if !v.HasErrors() {
			t.Error("Expected error for invalid hex color")
		}
	})

	t.Run("Valid hex color", func(t *testing.T) {
		v := NewValidator()
		v.ColorHex("color", "#FF0000")
		if v.HasErrors() {
			t.Error("Expected no error for valid hex color")
		}
	})

	t.Run("OneOf validation", func(t *testing.T) {
		v := NewValidator()
		v.OneOf("status", "invalid", []string{"draft", "published", "archived"})
		if !v.HasErrors() {
			t.Error("Expected error for value not in allowed list")
		}
	})
}
